File: design_patterns/c++/src/base/Door.cpp
#include "../../headers//Door.h"

Door::Door(Room* from, Room* to) {
    _room1 = from;
    _room2 = to;
    _isOpen = true;
}

Door::Door(const Door & other) {
  _room1 = other._room1;
  _room2 = other._room2;
}

Room* Door::otherSideFrom(Room* from) {
    if (from == _room1 && _isOpen) {
        return _room2;
    }
    if (from == _room2 && _isOpen) {
        return _room1;
    }
    return 0;
}

Door* Door::setRoom(Room* from, Room* to) {
  _room1 = from;
  _room2 = to;
  return this;
}

Door* Door::clone() const {
  return new Door(*this);
}

File: design_patterns/java/src/base/Door.java
package base;

public class Door implements MapSite, Cloneable {
    private Room mRoom1, mRoom2;
    public Door(Room room1, Room room2) {
        mRoom1 = room1;
        mRoom2 = room2;
    }


    @Override
    public boolean enter () {
        return true;
    }

    public Room otherSideFrom(Room from) {
        if (from == null) return null;
        if (from == mRoom1) {
            return mRoom2;
        } else if (from == mRoom2) {
            return mRoom1;
        } else {
            return null;
        }
    }

    @Override
    public Door clone() {
        // 或者使用 super.clone();
        return new Door(this.mRoom1, this.mRoom2);
    }
}
